From 83afcef5cfc215cc59508a1e16b9d0f59592a826 Mon Sep 17 00:00:00 2001
From: Hung Thanh Huynh <hung.huynh.wz@renesas.com>
Date: Fri, 22 Sep 2023 11:19:37 +0900
Subject: [PATCH 11/17] omx: videoenc: Convert I420 to NV12 process for RZV2H

RZV2H do not support I420 format, so we change the OMX parameter to
YUV420SemiPlanar and convert data from I420 to NV12 format.
So that we can support I420 in userside.

Signed-off-by: Hung Thanh Huynh <hung.huynh.wz@renesas.com>
---
 omx/gstomxvideoenc.c | 128 +++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 122 insertions(+), 6 deletions(-)

diff --git a/omx/gstomxvideoenc.c b/omx/gstomxvideoenc.c
index db3c276..f14e729 100644
--- a/omx/gstomxvideoenc.c
+++ b/omx/gstomxvideoenc.c
@@ -56,6 +56,11 @@ GST_DEBUG_CATEGORY_STATIC (gst_omx_video_enc_debug_category);
 static gboolean gst_omx_video_enc_set_use_buffer (GstOMXVideoEnc * self);
 #endif
 
+#ifdef USE_TARGET_DEVICE_RZV2H
+static gboolean
+gst_omx_video_enc_convertI420toNV12 (GstOMXVideoEnc * self, GstVideoFrame * frame, GstOMXBuffer * outbuf);
+#endif
+
 #define GST_TYPE_OMX_VIDEO_ENC_CONTROL_RATE (gst_omx_video_enc_control_rate_get_type ())
 static GType
 gst_omx_video_enc_control_rate_get_type (void)
@@ -2616,8 +2621,10 @@ gst_omx_video_enc_set_format (GstVideoEncoder * encoder,
     /* Fallback */
     switch (info->finfo->format) {
       case GST_VIDEO_FORMAT_I420:
+#ifndef USE_TARGET_DEVICE_RZV2H
         port_def.format.video.eColorFormat = OMX_COLOR_FormatYUV420Planar;
         break;
+#endif
       case GST_VIDEO_FORMAT_NV12:
         port_def.format.video.eColorFormat = OMX_COLOR_FormatYUV420SemiPlanar;
         break;
@@ -2642,6 +2649,14 @@ gst_omx_video_enc_set_format (GstVideoEncoder * encoder,
 
       if (m->format == info->finfo->format) {
         port_def.format.video.eColorFormat = m->type;
+#ifdef USE_TARGET_DEVICE_RZV2H
+        /* RZV2H supported only NV12 */
+        if (OMX_COLOR_FormatYUV420Planar == port_def.format.video.eColorFormat) {
+          port_def.format.video.eColorFormat = OMX_COLOR_FormatYUV420SemiPlanar;
+        } else if (OMX_COLOR_FormatYUV420PlanarMultiPlane == port_def.format.video.eColorFormat) {
+          port_def.format.video.eColorFormat = OMX_COLOR_FormatYUV420SemiPlanarMultiPlane;
+        }
+#endif
         break;
       }
     }
@@ -2923,12 +2938,20 @@ gst_omx_video_enc_fill_buffer (GstOMXVideoEnc * self, GstBuffer * inbuf,
 
     GST_LOG_OBJECT (self, "Matched strides - direct copy %u bytes",
         (guint) outbuf->omx_buf->nFilledLen);
-
-    gst_buffer_extract (inbuf, 0,
-        outbuf->omx_buf->pBuffer + outbuf->omx_buf->nOffset,
-        outbuf->omx_buf->nFilledLen);
-    ret = TRUE;
-    goto unmap_frame;
+#ifdef USE_TARGET_DEVICE_RZV2H
+    if (GST_VIDEO_FORMAT_I420 == info->finfo->format) {
+      ret = gst_omx_video_enc_convertI420toNV12 (self, &frame, outbuf);
+      if (FALSE == ret)
+        goto unmap_frame;
+    } else
+#endif
+    {
+      gst_buffer_extract (inbuf, 0,
+          outbuf->omx_buf->pBuffer + outbuf->omx_buf->nOffset,
+          outbuf->omx_buf->nFilledLen);
+      ret = TRUE;
+      goto unmap_frame;
+    }
   }
 
   /* Different strides */
@@ -2936,6 +2959,11 @@ gst_omx_video_enc_fill_buffer (GstOMXVideoEnc * self, GstBuffer * inbuf,
 
   switch (info->finfo->format) {
     case GST_VIDEO_FORMAT_I420:{
+#ifdef USE_TARGET_DEVICE_RZV2H
+      ret = gst_omx_video_enc_convertI420toNV12 (self, &frame, outbuf);
+      if (FALSE == ret)
+        goto unmap_frame;
+#else
       gint i, j, height, width;
       guint8 *src, *dest;
       gint src_stride, dest_stride;
@@ -2991,6 +3019,7 @@ gst_omx_video_enc_fill_buffer (GstOMXVideoEnc * self, GstBuffer * inbuf,
               (port_def->format.video.nSliceHeight / 2) *
               (port_def->format.video.nStride / 2);
       }
+#endif
       ret = TRUE;
       break;
     }
@@ -3711,3 +3740,90 @@ gst_omx_video_enc_decide_allocation (GstVideoEncoder * encoder,
 
   return TRUE;
 }
+
+#ifdef USE_TARGET_DEVICE_RZV2H
+static gboolean
+gst_omx_video_enc_convertI420toNV12 (GstOMXVideoEnc * self, GstVideoFrame * frame, GstOMXBuffer * outbuf)
+{
+  /* RZV2M supported only NV12 */
+  GstVideoCodecState *state = gst_video_codec_state_ref (self->input_state);
+  GstVideoInfo *info = &state->info;
+  OMX_PARAM_PORTDEFINITIONTYPE *port_def = &self->enc_in_port->port_def;
+  gboolean ret = TRUE;
+  gint i, j, k, height, width;
+  guint8 *srcY, *srcU, *srcV, *dest;
+  gint src_stride, dest_stride;
+
+  outbuf->omx_buf->nFilledLen = 0;
+
+  dest_stride = GST_ROUND_UP_128(port_def->format.video.nFrameWidth);
+
+  for (i = 0; i < 2; i++) {
+    src_stride = GST_VIDEO_FRAME_COMP_STRIDE (frame, i);
+    /* XXX: Try this if no stride was set */
+    if (dest_stride == 0)
+      dest_stride = src_stride;
+
+    dest = outbuf->omx_buf->pBuffer + outbuf->omx_buf->nOffset;
+    if (i == 0) {
+      srcY = GST_VIDEO_FRAME_COMP_DATA (frame, i);
+      height = GST_VIDEO_FRAME_COMP_HEIGHT (frame, i);
+      width = GST_VIDEO_FRAME_COMP_WIDTH (frame, i);
+
+      if (GST_VIDEO_FORMAT_INFO_BITS (info->finfo) == 10)
+        /* Need ((width + 2) / 3) 32-bits words */
+        width = (width + 2) / 3 * 4;
+
+      if (dest + dest_stride * height >
+          outbuf->omx_buf->pBuffer + outbuf->omx_buf->nAllocLen) {
+        GST_ERROR_OBJECT (self, "Invalid output buffer size");
+        ret = FALSE;
+        break;
+      }
+
+      for (j = 0; j < height; j++) {
+        memcpy (dest, srcY, width);
+        srcY += src_stride;
+        dest += dest_stride;
+      }
+
+      /* nFilledLen should include the vertical padding in each slice (spec 3.1.3.7.1) */
+      outbuf->omx_buf->nFilledLen +=
+          port_def->format.video.nSliceHeight *
+          port_def->format.video.nStride;
+    } else {
+      dest += height * dest_stride;
+      srcU = GST_VIDEO_FRAME_COMP_DATA (frame, i);
+      srcV = GST_VIDEO_FRAME_COMP_DATA (frame, i+1);
+      height = height / 2;
+      width = width / 2;
+      if (dest + dest_stride * height >
+          outbuf->omx_buf->pBuffer + outbuf->omx_buf->nAllocLen) {
+        GST_ERROR_OBJECT (self, "Invalid output buffer size");
+        ret = FALSE;
+        break;
+      }
+
+      for (j = 0; j < height; j++) {
+        for (k = 0; k < width; k++) {
+          dest[0] = *srcU;
+          dest[1] = *srcV;
+          srcU++;
+          srcV++;
+          dest += 2;
+        }
+        srcU += src_stride - width;
+        srcV += src_stride - width;
+        dest += dest_stride - (width * 2);
+      }
+
+      /* nFilledLen should include the vertical padding in each slice (spec 3.1.3.7.1) */
+      outbuf->omx_buf->nFilledLen +=
+          port_def->format.video.nSliceHeight / 2 *
+          port_def->format.video.nStride;
+    }
+  }
+
+  return ret;
+}
+#endif
-- 
2.25.1

